### PR TITLE
feat(Pipeline): `optimize()`

### PR DESCRIPTION
**Pull request by eddiebergman**
TODO:
* Cleaning up documentation errors
* Creating a issue to re-write some of the optimization guide which should draw heavy attention to this feature.

Alright, buckle up, this is a bigger one (780 lines added).

The objective of this PR was to enable `pipeline.optimize(...)` to support fastest defaults towards just, well evaluating a pipeline, while still enabling it to extend out further into the future and not be too locked up in terms of use case.

### Get me HPO results quickly
Let's start with the simplest procedure:

```python
pipline = ...
history = pipeline.optimize(...)  # <- What needs to go in here
```

The first thing is the `.optimizer(target=...)` which should be a function that given a `Trial` and `Node` (pipeline), should return a `Trial.Report`, formally: `Callable[[Trial, Node], Trial.Report]`. This is the users target function to use HPO nomeclature.

The `target=` can also be an `EvaluationProtocol`, which will be used to define pre-defined flows so to speak. For example, a future addition would allow something like:

```python
pipeline.optimize(
    target=SklearnCVProtocol(X, y, splitter=...),  # User can use predefined evaluation protocols
    ...
)
```

### Choosing an Optimizer and Creating one
The next parts revolve around the optimizer, namely `optimizer=` to select one, the `seed=`, `working_dir=` and `metrics=` it will expect. These constitue the changes to the `Optimizer` class. By default, it's `None` and just tries to get some optimizer installed and use that. There could be extra work to also detect what's compatible with the pipeline, but if they're using a custom seearch space defintion, then they likely will be able to pass in the specific optimizer they wanted anywho.

### Scheduling
Next things to consider were about task execution, namely around the `Scheduler` and running it. It just uses `n_workers: int = 1` and `scheduler: Scheduler | None = None` to mean that it runs the opimization in local processes with `1` worker. Both of these parameters can be freely changed though. There are also a host of parameters related to error control, plugins and `max_trials=` and more which are documented in the code and understandable through there.

### Finer Control
The last major part to understand is `setup_only: bool = False`. We assume the default behaviour is as the code snippet above, "just get me some HPO runs". However more advanced users (me) might want to have multiple pipelines setup to run HPO, but I do not want to start them yet, and only care more about setting up the entire flow.

```python
pipeline = ...
scheduler = pipeline.optimizer(..., setup_only=True)
```

In this case, the pipeline is ready to be optimized and will run as soon as I call `scheduler.run()` where notably, I am now in control of the scheduler and the `run()` call.

---

Other use cases considered:
* You want to use your own `History` object which may be shared between pipeline optimization runs. `optimize(history=my_history)`. This won't lock or cause race conflicts as all callbacks happen async in the main process.
* You may want custom callbacks, this is handled through `on_begin` which allows you to do a host of custom things which have been documented.

### CODE REVIEW

File: src/amltk/optimization/optimizer.py [hunk 1]
```diff
@@ -123,3 +130,92 @@ def preferred_parser(
...
+        metrics: Metric | Sequence[Metric],
+        bucket: str | Path | PathBucket | None = None,
+        seed: Seed | None = None,
+    ) -> Self:
+        """Create this optimizer.
+
+        !!! note
+
+            Subclasses should override this with more specific configuration
+            but these 3 arguments should be all that's necessary to create
```
Comment (eddiebergman) [inline src/amltk/optimization/optimizer.py:149]: Should be `4` or better yet, don't count

File: src/amltk/optimization/optimizer.py [hunk 2]
```diff
@@ -123,3 +130,92 @@ def preferred_parser(
...
+            space: The space to optimize over.
+            bucket: The bucket for where to store things related to the trial.
+            metrics: The metrics to optimize.
+            seed: The seed to use for the optimizer.
+
+        Returns:
+            The optimizer.
+        """
+
+    class CreateSignature(Protocol):
```
Comment (eddiebergman) [inline src/amltk/optimization/optimizer.py:162]: Note it's a `Protocol`, akin to a Java interface. I use it for type hinting what a function signature should look like, i.e. any function that looks like `Optimizer.create`

File: src/amltk/_util.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+from __future__ import annotations
+
+from typing import Any
+
+
+def threadpoolctl_heuristic(thing: Any | None) -> bool:
```
Comment (LennartPurucker) [inline src/amltk/_util.py:6]: Strange function IMO, and I do not understand it from the docstring.
Maybe add more text to explain the idea here or keep it this way but I would not be able to use it right now.

File: src/amltk/pipeline/node.py [hunk 4]
```diff
@@ -776,3 +865,508 @@ def _fufill_param_requests(
                     continue
 
         return new_config
+
+    @overload
```
Comment (LennartPurucker) [inline src/amltk/pipeline/node.py:869]: Mhm, I really do not like overload anymore after seeing this. 
Do you know if Python plans to move on or get some better syntax for this? Or does something better exist that does the same (for mypy)?

File: src/amltk/pipeline/node.py [hunk 5]
```diff
@@ -776,3 +865,508 @@ def _fufill_param_requests(
...
+            )
+            plugins = (*_plugins, plugin)
+
+        # If threadpool_limit_ctl None, we should default to inspecting if it's
+        # an sklearn pipeline. This is because sklearn pipelines
+        # run in parallel will over-subscribe the CPU and cause
+        # the system to slow down.
+        # We use a heuristic to check this by checking if the item at the head
+        # of this node is a subclass of sklearn.base.BaseEstimator
+        match threadpool_limit_ctl:
```
Comment (LennartPurucker) [inline src/amltk/pipeline/node.py:1185]: I swear, you just wrote this function because you wanted to use match cases :D

**Review by LennartPurucker — CHANGES_REQUESTED** [review]
Minor points, but one big one: does this have no tests?

File: src/amltk/pipeline/node.py (hunk 5 above)
Comment (eddiebergman) [reply src/amltk/pipeline/node.py:1185]: You should see some tool I'm building with AMLTK ;) Most of it is just parsing user arguments and if it doesn't match anything throw an error. Still better than a truck ton of `if`s :)

File: src/amltk/_util.py (hunk 3 above)
Comment (LennartPurucker) [reply src/amltk/_util.py:6]: Oh, I think what you are doing is great (now that I understand). I think simply adding to the docstring that the function is used for something like this would be great. I did not really understand what `thing` as input meant here, but if it is basically just a check if threadpoolctl is correctly configured for sklearn pipelines, then it's good.

**Comment by eddiebergman** [comment]
Not yet, I wanted to flesh it out properly first before testing it. Also the docs were taking to long to render so I have a PR that fixes that first, coming up

**Comment by eddiebergman** [comment]
Any advice on what to test would be appreciated. It's the weird kind of function which merely pieces things together but doesn't implement much, i.e. I'd be mainly testing that the matches do indeed match, the one thing concrete would be the heuristic

**Review by aron-bram — COMMENTED** [review]
Easy to read and clean!
I don't actually see what you would need to test here to be honest, assuming that the classes in use here are already tested individually, which they seem to be.

File: src/amltk/evalutors/evaluation_protocol.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+from amltk.scheduling import Plugin
+
+if TYPE_CHECKING:
+    from amltk.optimization import Trial
+    from amltk.pipeline import Node
+    from amltk.scheduling import Scheduler, Task
+
+
+class EvaluationProtocol:
+    """A protocol for how a trial and a pipeline should be evaluated."""
```
[the remark this answers is not in the document]
Comment (eddiebergman) [reply src/amltk/evalutors/evaluation_protocol.py:19]: Yup you're right. I don't have concrete implementations yet so it may change. You can imagine one implementation would be `SklearnCVEvaluator` which does a typical sklearn evaluation

File: src/amltk/evalutors/evaluation_protocol.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            case Iterable():
+                _plugins = tuple(plugins)
+
+        return scheduler.task(self.fn, plugins=_plugins)
+
+
+class CustomProtocol(EvaluationProtocol):
+    """A custom evaluation protocol based on a user function."""
+
+    def __init__(self, fn: Callable[[Trial, Node], Trial.Report]) -> None:
```
[the remark this answers is not in the document]
Comment (eddiebergman) [reply src/amltk/evalutors/evaluation_protocol.py:52]: The idea is that I don't really care how you create an `EvaluationProtocol`, just that I can get a `Task` from them with a specific signature.

File: src/amltk/optimization/optimizer.py [hunk 8]
```diff
@@ -123,3 +130,92 @@ def preferred_parser(
...
+        Args:
+            space: The space to optimize over.
+            bucket: The bucket for where to store things related to the trial.
+            metrics: The metrics to optimize.
+            seed: The seed to use for the optimizer.
+
+        Returns:
+            The optimizer.
+        """
+
```
[the remark this answers is not in the document]
Comment (eddiebergman) [reply src/amltk/optimization/optimizer.py:161]: Yup, technically, as long as there's something in the body (i.e. a docstring) then it's a valid function definition. The `...` can be useful when it doesn't make sense to add a docstring and you need _something_ in the body. Technically it's the same as if I did:

```python
def f():
    4
```

The `...` is a value that you can even pass around of type `_EllipsesType` ... I once had a weird function where this was a value to pass in.


```python
def reorder_columns(df: pd.DataFrame, columns: list[str | _ElipsesType]):
     """The elipses means 'rest'"""
     
reorder_columns(my_df, ["col42", "col_other", ...])
```

File: src/amltk/evalutors/evaluation_protocol.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+        Args:
+            scheduler: The scheduler to use for the task.
+            plugins: The plugins to use for the task.
+
+        Returns:
+            The created task.
+        """
+        _plugins: tuple[Plugin, ...]
+        match plugins:
```
[the remark this answers is not in the document]
Comment (aron-bram) [reply src/amltk/evalutors/evaluation_protocol.py:38]: Great, I will start using it :slightly_smiling_face:

File: src/amltk/evalutors/evaluation_protocol.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (aron-bram) [reply src/amltk/evalutors/evaluation_protocol.py:52]: Oh, okay, so it just serves as an example for the users.

[the remark this answers is not in the document]
Comment (eddiebergman) [reply src/amltk/evalutors/evaluation_protocol.py:52]: Nay, it's still useful to get user provided callables. There'll be more documentation once I've created an actual useful one

**Comment by eddiebergman** [comment]
Thank you for the reviews @LennartPurucker @aron-bram! I made changes based on your comments.

Feel free to review post-mortem for any issues but after adding some tests and verifying it seems to be working as intended.

Note: Ignore what I said about testing... I found small variable name bugs which caused behaviors not to occur. Live by the tests, die by the tests

File: src/amltk/optimization/optimizer.py [hunk 10]
```diff
@@ -146,8 +146,7 @@ def create(
         !!! note
 
             Subclasses should override this with more specific configuration
-            but these 3 arguments should be all that's necessary to create
-            the optimizer.
+            but these arguments should be all that's necessary to create the optimizer.
```
Comment (aron-bram) [inline src/amltk/optimization/optimizer.py:149]: Future proofing docs :slightly_smiling_face:

File: src/amltk/pipeline/node.py [hunk 11]
```diff
@@ -1083,7 +1083,7 @@ def register_optimization_loop(  # noqa: C901, PLR0915, PLR0912
                 walltime_limit=process_walltime_limit,
                 cputime_limit=process_cputime_limit,
             )
-            plugins = (*_plugins, plugin)
+            _plugins = (*_plugins, plugin)
```
Comment (aron-bram) [inline src/amltk/pipeline/node.py:1086]: I also tend to easily make these type of mistakes...
